core/src/broadcast_stage/broadcast_different_blobs_run.rs

```diff
@@ -0,0 +1,96 @@
+use super::*;
+use crate::entry::Entry;
+use solana_sdk::hash::Hash;
+
+pub(super) struct BroadcastDifferentBlobsRun {
+    last_blockhash: Hash,
+}
+
+impl BroadcastDifferentBlobsRun {
+    pub(super) fn new() -> Self {
+        Self {
+            last_blockhash: Hash::default(),
+        }
+    }
+}
+
+impl BroadcastRun for BroadcastDifferentBlobsRun {
+    fn run(
+        &mut self,
+        broadcast: &mut Broadcast,
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        receiver: &Receiver<WorkingBankEntries>,
+        sock: &UdpSocket,
+        blocktree: &Arc<Blocktree>,
+    ) -> Result<()> {
+        // 1) Pull entries from banking stage
+        let receive_results = broadcast_utils::recv_slot_blobs(receiver)?;
+        let bank = receive_results.bank.clone();
+        let last_tick = receive_results.last_tick;
+
+        let keypair = &cluster_info.read().unwrap().keypair.clone();
+        let latest_blob_index = blocktree
+            .meta(bank.slot())
+            .expect("Database error")
+            .map(|meta| meta.consumed)
+            .unwrap_or(0);
+
+        let (data_blobs, coding_blobs) = broadcast_utils::entries_to_blobs(
+            receive_results.ventries,
+            &broadcast.thread_pool,
+            latest_blob_index,
+            last_tick,
+            &bank,
+            &keypair,
+            &mut broadcast.coding_generator,
+        );
+
+        // If the last blockhash is default, a new block is being created
+        // So grab the last blockhash from the parent bank
+        if self.last_blockhash == Hash::default() {
+            self.last_blockhash = bank.parent().unwrap().last_blockhash();
+        }
+
+        let fake_ventries: Vec<_> = (0..receive_results.num_entries)
+            .map(|_| vec![(Entry::new(&self.last_blockhash, 0, vec![]), 0)])
+            .collect();
+
+        let (fake_data_blobs, fake_coding_blobs) = broadcast_utils::entries_to_blobs(
+            fake_ventries,
+            &broadcast.thread_pool,
+            latest_blob_index,
+            last_tick,
+            &bank,
+            &keypair,
+            &mut broadcast.coding_generator,
+        );
+
+        // If it's the last tick, reset the last block hash to default
+        // this will cause next run to grab last bank's blockhash
+        if last_tick == bank.max_tick_height() {
+            self.last_blockhash = Hash::default();
+        }
+
+        blocktree.write_shared_blobs(data_blobs.iter().chain(coding_blobs.iter()))?;
+
+        // 3) Start broadcast step
+        let bank_epoch = bank.get_stakers_epoch(bank.slot());
+        let stakes = staking_utils::staked_nodes_at_epoch(&bank, bank_epoch);
+
+        // Broadcast data + erasures
```

Comment:
I'm guessing what will happen due to turbine if we broadcast the two sets of blobs disjointly like this is that most nodes will get a mix of blobs from set 1 and set 2 (especially since we're not broadcasting them all at once). This would test a more extreme, but similar behavior as the FailEntryVerificationBroadcastRun.

I think a unique test case  that the other broadcast run doesn't cover would be what happens if a node first gets all of the blobs from one of these sets (so the entry verification doesn't fail in the middle of a slot), and then he sees blobs from the other set. Maybe this can be achieved by sleeping for a bit before broadcasting the second set of blobs? I think it would also mean we would have to broadcast all the blobs at once.

Reply:
👍